text/0000-platform-default-port.md

```diff
@@ -0,0 +1,50 @@
+# Meta
+[meta]: #meta
+- Name: Default ports
+- Start Date: 2020-01-22
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+Proposes a mechanism whereby a platform can expose a default port for builds.
+
+# Motivation
+[motivation]: #motivation
+
+A number of users have complained that the containers they created with CNBs don't run "out of the box" because of an unset `$PORT` variable. This leads to a confusing initial experience and makes the path from running a simple `pack build` to starting a container unnecessarily rocky. If platforms had a default value of `$PORT` they could expose to the lifecycle, which could easily be overridden during a build, it would reduce some of that friction.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Platforms that build images with CNB tech create images with a default value for `PORT=8080`. Buildpacks can override that value if it makes sens for their individual case; for example a `postgres` buildpack - if there were such a thing - would override `$PORT` with `5432` which a common convention. In cases where there is no assumed value because the language community hasn't standardized (for example in ruby or nodejs) the images produced by, for example, pack, could still start without the added step of configuring this value during invocation.
```

Comment:
As a developer I would expect the opposite behaviour. I would expect that there is a default set by the buildpack that I could then override via parameter at build time. In most cases I believe env vars is the way to communicate configuration to buildpacks.

Essentially what I picture is an order of precedence:
1. Environment variables
2. Buildpack defaults

For example:
1. `pack build ... -env POSGRES_PORT=6543 ...`
2. buildpack defaults to `5432` if `POSGRES_PORT` is not present

Maybe this is simply a question of buildpack best practices such as informing the user of the port being bound and what they can do to override it OR there could be a better mechanism for discovering configurability within buildpacks.

Reply:
What if..., hear me out, we had a standard way to declare configuration within `buildpack.toml` that platforms could then use to display configuration elements.

For example a `buildpack.toml`:
```toml
[buildpack]
id = "..."
version = "..."

[config]
[[config.env]]
name = "POSTGRES_PORT"
description="Port to serve PostgreSQL on."
default="5432"
```

`pack` could then do various things with that information such as:

```shell script
$ pack get-buildpack-config postgres@1.2.3

Environment Variables:
  POSTGRES_PORT        Port to serve PostgreSQL on. (default 5432)

```

Reply:
🌈 This discussion will be moved to a seperate RFC.